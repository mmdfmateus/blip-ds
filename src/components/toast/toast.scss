@mixin inherit-color() {
  color: inherit;
  background-color: inherit;
}
:host {
  .show, .hide {
    display: flex;
  }
  .show {
    opacity: 1;

    &--top-right,
    &--bottom-right {
      animation: toastAnimationFadeInFromRight 1s;
    }
    
    &--top-left,
    &--bottom-left {
      animation: toastAnimationFadeInFromLeft 1s;
    }
  }
  .hide {
    transition: all 1s;
    animation: toastAnimationFadeOut 0.5s;
  }
}

.toast {
  display: none;
  position: relative;
  align-items: center;
  box-sizing: border-box;
  border-radius: 8px;
  box-shadow: $shadow-toast;
  color: $color-snow-white;
  opacity: 0;
  margin-top: 16px;

  &--action--icon {
    min-width: 440px;
    max-width: 440px;
    padding: 8px 20px;
    
    bds-icon-button {
      height: 32px;
    }

    @media (max-width: $sm-screen) {
      min-width: 220px;
      width: 95%;
      margin: 16px auto 0px auto;
    }
  }

  &--action--button {
    min-width: 440px;
    max-width: 456px;
    padding: 10px 20px;

    @media (max-width: $sm-screen) {
      min-width: 220px;
      width: 95%;
      margin: 16px auto 0px auto;
    }
  }

  &--system {
    background: $color-primary-main;
  }

  &--error {
    background: $color-secondary-delete;
  }

  &--success {
    background: $color-secondary-true;
  }

  &--warning {
    background: $color-secondary-cheetos;
  }

  &--undo{
    background-color: $color-primary-main;
  }

  &--redo{
    background-color: $color-secondary-redo;
  }

  bds-icon {
    margin-right: 8px;
    display: flex;
    align-items: center;
  }

  &__content {
    height: 100%;
    width: 100%;
    align-items: flex-start;
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-right: 8px;
  }

  &__action {
    display: flex;
    align-items: center;

    button {
      background: transparent;
      border: none;
      outline: none;
      padding: 0;
    }

    &__button {
      white-space: nowrap;
    }

    &__icon {
      display: flex !important;
      align-items: center;

      bds-icon {
        color: $color-snow-white;
        border-radius: 4px;
        cursor: pointer;
      }
    }
  }
}

@keyframes toastAnimationFadeInFromRight {
  0% {
    opacity: 0;
    right: -200px;
  }
  50% {
    opacity: 0.9;
    right: 1px;
  }
  100% {
    opacity: 1;
  }
}

@keyframes toastAnimationFadeInFromLeft {
  0% {
    opacity: 0;
    left: -200px;
  }
  50% {
    opacity: 0.9;
    left: 1px;
  }
  100% {
    opacity: 1;
  }
}

@keyframes toastAnimationFadeOut {
  0% {
    opacity: 1;
  }
  30% {
    max-height: 60px;
  }
  80% {
    opacity: 0;
    max-height: 30px;
  }
  100% {
    max-height: 0px;
  }
}
