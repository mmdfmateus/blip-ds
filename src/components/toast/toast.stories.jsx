import React from "react";
import { withKnobs, text, select } from "@storybook/addon-knobs";

import readme from "./readme.md";

export default {
  title: "Toast",
  decorators: [withKnobs],
  parameters: {
    notes: { markdown: readme },
  },
};

export const toastSystem = () => {
  const variantOptions = {
    System: "system",
    Error: "error",
    Success: "success",
    Warning: "warning",
    Undo: "undo",
    Redo: "redo"
  };
  const actionTypeOptions = { Button: "button", Icon: "icon" };
  const buttonActionOptions = { Close: "close", Custom: "custom" };
  const positionTypeOptions = { BottomLeft: 'bottom-left', BottomRight: 'bottom-right', TopLeft: 'top-left', TopRight: 'top-right' };

  const showToast = async () => {
    const newtoast = document.createElement("bds-toast");
    document.body.appendChild(newtoast);

    const icon = text("icon", undefined);
    const variant = select("variant", variantOptions);
    const actionType = select("action-type", actionTypeOptions);
    const toastTitle = text("toast-title", "Toast Title");
    const toastText = text("toast-text", "Lorem ipsum");
    const buttonText = text("button-text", "Cancelar");
    const duration = text("duration", 0);
    const buttonAction = select("button-action", buttonActionOptions);
    const position = select("position", positionTypeOptions, positionTypeOptions.BottomLeft);

    await newtoast.create({
      variant,
      actionType,
      toastTitle,
      buttonText,
      buttonAction,
      toastText,
      variant,
      duration,
      icon,
      position
    });
  };

  return (
    <>
      <bds-button onClick={() => showToast()}>Show toast</bds-button>
    </>
  );
};
