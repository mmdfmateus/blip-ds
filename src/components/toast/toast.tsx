import { Component, ComponentInterface, h, Prop, Method, Element, Event, EventEmitter } from '@stencil/core';
import {
  ActionType,
  VariantType,
  ButtonActionType,
  CreateToastType,
  IconVariantMap,
  PositionType,
} from './toast-interface';
@Component({
  tag: 'bds-toast',
  styleUrl: 'toast.scss',
  shadow: true,
})
export class BdsToast implements ComponentInterface {
  @Element() el: HTMLBdsToastElement;
  /**
   * used for add the icon. Uses the bds-icon component.
   */
  @Prop({ reflect: true }) icon?: string = null;
  /**
   * ActionType. Defines if the button should have a button or an icon. Can be one of:
   * 'icon', 'button';
   */
  @Prop() actionType: ActionType = 'button';
  /**
   * Variant. Defines the color of the toast. Can be one of:
   * 'system', 'error', 'success', 'warning', 'undo', 'redo';
   */
  @Prop() variant: VariantType = 'system';
  /**
   * The title of the component:
   */
  @Prop() toastTitle: string;
  /**
   * The text content of the component:
   */
  @Prop() toastText: string;
  /**
   * If the action type is button, this will be the text of the button:
   */
  @Prop() buttonText: string;
  /**
   * Time to close the toast in seconds
   * 0 = never close automatically (default value)
   */
  @Prop() duration = 0;
  /**
   * Define an action to the button toast. Can be one of:
   * 'close', 'custom';
   * if the action type is set to close, the button will close automatically.
   * if the action type is set to custom, a function need to be passed when the toastButtonClick is emitted.
   */
  @Prop() buttonAction: ButtonActionType = 'close';
  /**
   * Controls the open event of the component:
   */
  @Prop() show = false;
  /**
   * Controls the hide event of the component:
   */
  @Prop() hide = false;
  /**
   * The toast position on the screen. Can be one of:
   * 'top-right', 'top-left', 'bottom-right', 'bottom-left' (default value);
   */
  @Prop() position: PositionType = 'bottom-left';
  /**
   * Event used to execute some action when the action button on the toast is clicked
   */
  @Event() toastButtonClick!: EventEmitter;
  /**
   * Sends an event to be used when creating an action when clicking the toast button
   */
  private _buttonClickHandler = () => {
    if (this.buttonAction === 'close') this.close();
    else {
      this.toastButtonClick.emit(this.el);
      this.close();
    }
  };

  /**
   * Can be used outside to open the toast
   */
  @Method()
  async create({
    actionType,
    buttonAction,
    buttonText,
    icon,
    toastText,
    toastTitle,
    variant,
    duration,
    position,
  }: CreateToastType) {
    let toastContainer = document.querySelector(`bds-toast-container.${position}`);

    if (toastContainer) {
      toastContainer.appendChild(this.el);
      toastContainer.classList.add(position);
    } else {
      toastContainer = document.createElement('bds-toast-container');
      toastContainer.classList.add(position);
      document.body.appendChild(toastContainer);
      toastContainer.appendChild(this.el);
    }
    this.el.actionType = actionType || 'button';
    this.el.buttonAction = buttonAction || 'close';
    this.el.buttonText = buttonText;
    this.el.toastText = toastText;
    this.el.toastTitle = toastTitle;
    this.el.variant = variant || 'system';
    this.el.duration = duration * 1000 || 0;
    this.el.position = position || 'bottom-left';

    this.el.icon = icon ?? this.mapIconName[this.variant];

    this.el.show = true;

    if (this.el.duration > 0) {
      setTimeout(() => {
        this.el.hide = true;
        setTimeout(() => {
          this.el.remove();
        }, 400);
      }, this.el.duration);
    }
  }

  /**
   * Can be used outside the component to close the toast
   */
  @Method()
  async close() {
    if (this.el.shadowRoot) {
      this.el.shadowRoot.querySelector('div').classList.remove('show');
      this.el.shadowRoot.querySelector('div').classList.add('hide');
    } else {
      this.el.querySelector('div').classList.remove('show');
      this.el.querySelector('div').classList.add('hide');
    }

    setTimeout(() => {
      this.el.remove();
    }, 400);
  }

  private mapIconName: IconVariantMap = {
    system: 'bell',
    error: 'error',
    success: 'like',
    warning: 'attention',
    undo: 'undo-arrow',
    redo: 'redo-arrow',
  };

  render() {
    return (
      <div
        class={{
          toast: true,
          [`toast--${this.variant}`]: true,
          [`toast--action--${this.actionType}`]: true,
          [`show show--${this.position}`]: this.show,
          hide: this.hide,
        }}
      >
        {this.icon && <bds-icon theme="outline" size="x-large" color="#fff" name={this.icon} />}
        <div class="toast__content">
          {this.toastTitle && (
            <bds-typo variant="fs-14" bold="bold">
              {this.toastTitle}
            </bds-typo>
          )}
          {this.toastText && <bds-typo variant="fs-14" innerHTML={this.toastText}></bds-typo>}
        </div>
        <div
          class={{
            toast__action: true,
            [`toast__action__${this.actionType}`]: true,
          }}
        >
          {this.actionType === 'button' ? (
            <button onClick={() => this._buttonClickHandler()}>
              <bds-button variant="secondary--white" size="short">
                {this.buttonText}
              </bds-button>
            </button>
          ) : (
            <button onClick={() => this._buttonClickHandler()}>
              <bds-icon-button size="short" variant="secondary--white" icon="close" />
            </button>
          )}
        </div>
      </div>
    );
  }
}
