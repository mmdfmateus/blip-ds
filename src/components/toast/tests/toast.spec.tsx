import { newSpecPage } from '@stencil/core/testing';
import { BdsToast } from '../toast';

describe('bds-toast', () => {
  const TOAST_REMOVED_TIMEOUT = 400;
  jest.useFakeTimers();

  const getToastWithoutShadowDom = async () => {
    return await newSpecPage({
      components: [BdsToast],
      html: `
      <bds-toast toast-title="titulo" toast-text="texto" action-type="button" button-action="close" button-text="cancel" variant="system" duration="0" show="true"></bds-toast>`,
      supportsShadowDom: false,
    });
  };

  const getToast = async () => {
    return await newSpecPage({
      components: [BdsToast],
      html: `
      <bds-toast toast-title="titulo" toast-text="texto" action-type="button" button-action="close" button-text="cancel" variant="system" duration="0" show="true"></bds-toast>`,
    });
  };

  const getToastCustomAction = async () => {
    return await newSpecPage({
      components: [BdsToast],
      html: `
      <bds-toast toast-title="titulo" toast-text="texto" icon="close" action-type="button" button-action="custom" button-text="cancel" variant="system" duration="0" show="true"></bds-toast>`,
    });
  };

  const getToastIconActionType = async () => {
    return await newSpecPage({
      components: [BdsToast],
      html: `
      <bds-toast toast-title="titulo" toast-text="texto" icon="true" action-type="icon" button-action="close" button-text="cancel" variant="system" duration="0" show="true"></bds-toast>`,
    });
  };

  const getEmptyToast = async () => {
    return await newSpecPage({
      components: [BdsToast],
      html: `
      <bds-toast></bds-toast>`,
    });
  };

  it('should build', () => {
    expect(new BdsToast()).toBeTruthy();
  });

  it('should render the component and match the snapshot', async () => {
    const page = await getToast();
    expect(page).toMatchSnapshot();
  });

  it('should render the component in the shadow dom', async () => {
    const page = await getToast();
    expect(page.root.shadowRoot.querySelector('.toast')).toBeTruthy();
    expect(page.root.querySelector('.toast')).toBeFalsy();
  });

  it('should render without shadow dom', async () => {
    const page = await getToastWithoutShadowDom();
    expect(page.root.querySelector('.toast')).toBeTruthy();
    expect(page.root.shadowRoot).toBeFalsy();
  });

  it('should execute a custom action on button click', async () => {
    const page = await getToastCustomAction();
    const toast = page.body.querySelector('bds-toast');
    const _callback = jest.fn();
    page.doc.addEventListener('toastButtonClick', _callback);

    await page.waitForChanges();

    toast.shadowRoot.querySelector('button').click();

    await page.waitForChanges();

    expect(_callback).toHaveBeenCalled();
  });

  it('should close the toast clicking in the icon', async () => {
    jest.clearAllTimers();
    const page = await getToastIconActionType();
    const toast = page.body.querySelector('bds-toast');

    toast.shadowRoot.querySelector('button').click();

    await page.waitForChanges();

    expect(page.root.shadowRoot.querySelector('.toast').classList.contains('show')).toBe(false);
    expect(page.root.shadowRoot.querySelector('.toast').classList.contains('hide')).toBe(true);

    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), TOAST_REMOVED_TIMEOUT);

    setTimeout(() => {
      expect(page.body.querySelector('bds-toast')).toBeFalsy();
    }, TOAST_REMOVED_TIMEOUT);
    jest.runAllTimers();
  });

  it('should create a new toast', async () => {
    const page = await getEmptyToast();

    await page.root.create({
      actionType: 'button',
      buttonAction: 'close',
      buttonText: 'cancel',
      icon: 'trash',
      toastText: 'lorem ipsum is very cool',
      toastTitle: 'Best title ever',
      variant: 'warning',
      duration: 0,
    });

    await page.waitForChanges();

    let newToast = page.root.shadowRoot.querySelector('.toast--warning');
    expect(newToast).toBeTruthy();

    newToast = page.root.shadowRoot.querySelector('.toast--action--button');
    expect(newToast).toBeTruthy();

    const icon = page.root.shadowRoot.querySelector('.toast').firstElementChild;
    expect(icon.getAttribute('name')).toBe('trash');
  });

  it('should be able to set all the props', async () => {
    const toast = await getEmptyToast();

    toast.root.buttonAction = 'close';
    toast.root.actionType = 'button';
    toast.root.buttonText = 'Close';
    toast.root.icon = 'info';
    toast.root.toastText = 'Lorem ipsum';
    toast.root.toastTitle = 'Toast';
    toast.root.variant = 'error';
    toast.root.duration = 5;

    await toast.waitForChanges();

    expect(toast.root).toMatchSnapshot();
  });

  it('should create a new toast on right position', async () => {
    const positions = ['top-left', 'top-right', 'bottom-right', 'bottom-left'];

    for (const position in positions) {
      const page = await getEmptyToast();

      await page.root.create({
        actionType: 'button',
        buttonAction: 'close',
        buttonText: 'cancel',
        icon: 'trash',
        toastText: 'lorem ipsum is very cool',
        toastTitle: 'Best title ever',
        variant: 'system',
        duration: 0,
        position,
      });

      await page.waitForChanges();

      const newToast = page.root.shadowRoot.querySelector(`.show--${position}`);
      expect(newToast).toBeTruthy();

      const toastContainer = page.body.querySelector('bds-toast-container');
      expect(toastContainer).toHaveClass(position);
    }
  });

  it('should render the default icon (attention) if the variant is warning icon prop is not passed', async () => {
    const page = await getToast();

    await page.root.create({
      actionType: 'button',
      buttonAction: 'close',
      buttonText: 'cancel',
      toastText: 'lorem ipsum is very cool',
      toastTitle: 'Best title ever',
      variant: 'warning',
      duration: 0,
      position: 'bottom-left',
    });

    await page.waitForChanges();

    const icon = page.root.shadowRoot.querySelector('.toast').firstElementChild;
    expect(icon.getAttribute('name')).toBe('attention');
  });

  it('should render the default icon (like) if the variant is success icon prop is not passed', async () => {
    const page = await getToast();

    await page.root.create({
      actionType: 'button',
      buttonAction: 'close',
      buttonText: 'cancel',
      toastText: 'lorem ipsum is very cool',
      toastTitle: 'Best title ever',
      variant: 'success',
      duration: 0,
    });

    await page.waitForChanges();

    const icon = page.root.shadowRoot.querySelector('.toast').firstElementChild;
    expect(icon.getAttribute('name')).toBe('like');
  });
});
