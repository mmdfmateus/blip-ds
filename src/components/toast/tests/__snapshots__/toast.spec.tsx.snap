// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bds-toast should be able to set all the props 1`] = `
<bds-toast icon="info">
  <mock:shadow-root>
    <div class="toast toast--action--button toast--error">
      <bds-icon color="#fff" name="info" size="x-large" theme="outline"></bds-icon>
      <div class="toast__content">
        <bds-typo bold="bold" variant="fs-14">
          Toast
        </bds-typo>
        <bds-typo variant="fs-14">
          Lorem ipsum
        </bds-typo>
      </div>
      <div class="toast__action toast__action__button">
        <button>
          <bds-button size="short" variant="secondary--white">
            Close
          </bds-button>
        </button>
      </div>
    </div>
  </mock:shadow-root>
</bds-toast>
`;

exports[`bds-toast should render the component and match the snapshot 1`] = `
Object {
  "body": <body>
  <bds-toast action-type="button" button-action="close" button-text="cancel" duration="0" show="true" toast-text="texto" toast-title="titulo" variant="system">
    <mock:shadow-root>
      <div class="show show--bottom-left toast toast--action--button toast--system">
        <div class="toast__content">
          <bds-typo bold="bold" variant="fs-14">
            titulo
          </bds-typo>
          <bds-typo variant="fs-14">
            texto
          </bds-typo>
        </div>
        <div class="toast__action toast__action__button">
          <button>
            <bds-button size="short" variant="secondary--white">
              cancel
            </bds-button>
          </button>
        </div>
      </div>
    </mock:shadow-root>
  </bds-toast>
</body>,
  "build": Object {
    "allRenderFn": true,
    "appendChildSlotFix": false,
    "asyncLoading": true,
    "asyncQueue": true,
    "attachStyles": false,
    "cloneNodeFix": false,
    "cmpDidLoad": true,
    "cmpDidRender": true,
    "cmpDidUnload": true,
    "cmpDidUpdate": true,
    "cmpShouldUpdate": true,
    "cmpWillLoad": true,
    "cmpWillRender": true,
    "cmpWillUpdate": true,
    "connectedCallback": true,
    "constructableCSS": true,
    "cssAnnotations": false,
    "cssVarShim": true,
    "devTools": false,
    "disconnectedCallback": true,
    "dynamicImportShim": false,
    "element": true,
    "event": true,
    "hasRenderFn": true,
    "hostListener": true,
    "hostListenerTarget": true,
    "hostListenerTargetBody": true,
    "hostListenerTargetDocument": true,
    "hostListenerTargetParent": true,
    "hostListenerTargetWindow": true,
    "hotModuleReplacement": false,
    "hydrateClientSide": false,
    "hydrateServerSide": false,
    "hydratedAttribute": false,
    "hydratedClass": true,
    "initializeNextTick": true,
    "isDebug": false,
    "isDev": true,
    "isTesting": true,
    "lazyLoad": true,
    "lifecycle": true,
    "lifecycleDOMEvents": true,
    "member": true,
    "method": true,
    "mode": true,
    "observeAttribute": true,
    "profile": true,
    "prop": true,
    "propBoolean": true,
    "propMutable": true,
    "propNumber": true,
    "propString": true,
    "reflect": true,
    "safari10": false,
    "scoped": true,
    "scriptDataOpts": false,
    "shadowDelegatesFocus": true,
    "shadowDom": true,
    "shadowDomShim": false,
    "slot": true,
    "slotChildNodesFix": false,
    "slotRelocation": true,
    "state": true,
    "style": true,
    "svg": true,
    "taskQueue": true,
    "transformTagName": true,
    "updatable": true,
    "vdomAttribute": true,
    "vdomClass": true,
    "vdomFunctional": true,
    "vdomKey": true,
    "vdomListener": true,
    "vdomPropOrAttr": true,
    "vdomRef": true,
    "vdomRender": true,
    "vdomStyle": true,
    "vdomText": true,
    "vdomXlink": true,
    "watchCallback": true,
  },
  "doc": ,
  "flushLoadModule": [Function],
  "flushQueue": [Function],
  "setContent": [Function],
  "styles": Map {},
  "waitForChanges": [Function],
  "win": [MockWindow],
}
`;
